jujubundlelib/validate.py

```diff
@@ -0,0 +1,198 @@
+from __future__ import unicode_literals
+
+import collections
+import re
+
+import models
+import utils
+
+
+class BundleValidator(object):
+
+    def __init__(self, bundle):
+        self.bundle = bundle
+        self._errors = []
+
+    def add_error(self, error):
+        self._errors.append(error)
+
+    def errors(self):
+        return self._errors
+
+    def _is_legacy_bundle(self):
+        return utils.is_legacy_bundle(self.bundle)
+
+
+def validate(bundle):
+    validator = BundleValidator(bundle)
+    machine_ref_counts = {}
+    for machine in validator.bundle.get('machines', {}).keys():
+        machine_ref_counts[machine] = 0
+
+    if not valid_series(validator.bundle.get('series', '')):
+        validator.add_error(
+            'bundle has invalid series {}'.format(
+                validator.bundle.get('series', 'None')))
+    validate_machines(validator)
+    validate_services(validator,
+                      machine_ref_counts=machine_ref_counts)
+    validate_relations(validator)
+    validate_options(validator)
+
+    for machine, count in machine_ref_counts:
+        if count == 0:
+            validator.add_error(
+                'machine {} not referred to by a placement '
+                'directive'.format(machine))
+    return validator.errors()
+
+
+def valid_series(series):
+    valid_series = [
+        'oneiric',
+        'precise',
+        'quantal',
+        'raring',
+        'saucy',
+        'trusty',
+        'utopic',
+        'vivid',
+    ]
+    if series not in valid_series:
+        return False
+    return True
+
+
+def valid_constraints(constraints):
+    constraint_types = (
+        'arch',
+        'cpu-cores',
+        'mem',
+        'root-disk',
+        'container',
+        'cpu-power',
+        'tags',
+        'networks',
+        'instance-type',
+    )
+    if len(constraints):
+        constraints = re.split(r'\s+', constraints)
+        for constraint in constraints:
+            parts = constraint.split('=')
+            if (len(parts) != 2 or
+                    (len(parts[0]) < 1 and len(parts[1]) < 1) or
```

Comment:
I think that len(parts[0]) < 1 is redundant given that we check parts[0] not in constraint_types later.